lib/protein/client.ex

```diff
@@ -180,41 +186,53 @@ defmodule Protein.Client do
     request_buf = request_mod.encode(request_struct)
 
     result =
-      call_via_mock(request_buf, request_mod, response_mod, mock_mod)
-      || call_via_adapter(service_name, request_buf, transport_opts)
+      call_via_mock(request_buf, request_metadata, request_mod, response_mod, mock_mod)
+      || call_via_adapter(service_name, request_buf, request_metadata, transport_opts)
 
     case result do
-      {:ok, response_buf} ->
+      {:ok, response_buf, _response_metadata} ->
         {:ok, response_mod.decode(response_buf)}
-      {:error, errors} ->
+      {:error, errors, _response_metadata} ->
         {:error, errors}
     end
   end
 
   @doc false
-  def call!(request_struct, service_opts, transport_opts) do
+  def call!(request_struct, request_metadata, service_opts, transport_opts) do
     request_struct
-    |> call(service_opts, transport_opts)
+    |> call(request_metadata, service_opts, transport_opts)
     |> handle_non_failing_response()
   end
 
-  defp call_via_mock(request_buf, request_mod, response_mod, mock_mod) do
+  defp call_via_mock(request_buf, request_metadata, request_mod, response_mod, mock_mod) do
     if Utils.mocking_enabled?() do
-      Server.process_service(mock_mod, request_buf, request_mod, response_mod)
+      Server.process_service(mock_mod, request_buf, request_metadata, request_mod, response_mod)
     end
   rescue
     error -> raise TransportError, adapter: :mock, context: error
   end
 
-  defp call_via_adapter(service_name, request_buf, opts) do
+  defp call_via_adapter(service_name, request_buf, request_metadata, opts) do
     {adapter, adapter_opts} = Keyword.pop(opts, :adapter)
-    request_payload = RequestPayload.encode(service_name, request_buf)
+    request_payload = RequestPayload.encode(service_name, request_buf, request_metadata)
+
+    log_with_metadata(fn -> "Calling RPC service: #{service_name}" end, request_metadata)
```

Comment:
So in order to support 3rd party event logging, we could add `loggers` config key, either per-service (preferably but requires getting hands on & passing through `__service_opts__`) or global under `:protein` key. This would expect a module (or list of modules) that has `log(event)` function (may use behavior to enforce that). We'd bundle a default `Protein.LogEntry` logger that logs to console just like you do in-place now. May look into how Ecto models this for details.

Events would be structs:
- `Protein.ClientCallStartEvent` (service_name, request_metadata, timestamp)
- `Protein.ClientCallEndEvent` (start_event, response_metadata, timestamp)
- `Protein.ClientPushEvent` (service_name, request_metadata, timestamp)
- `Protein.ServerProcessingStartEvent` (service_name, request_metadata, timestamp)
- `Protein.ServerProcessingEndEvent` (start_event, response_metadata, timestamp)

I don't think errors need to be logged as in Elixir they're properly raised (it was an issue in `protein-ruby` to log them to Sentry).